chia/plotters/bladebit.py

```diff
@@ -0,0 +1,246 @@
+                capture_output=True,
+                text=True,
+            )
+            version = proc.stdout.strip()
+        except Exception as e:
+            print(f"Failed to determine bladebit version: {e}")
+
+        if version is not None:
+            installed = True
+            info["version"] = version
+        else:
+            installed = False
+
+    info["installed"] = installed
+    if installed is False:
+        info["can_install"] = supported
+
+    if supported and meets_memory_requirement() is False:
+        info["bladebit_memory_warning"] = f"BladeBit requires at least {BLADEBIT_MIN_RAM_GIBS} GiB of RAM to operate"
+
+    return info
+
+
+progress = {
+    "Finished F1 sort": 0.01,
+    "Finished forward propagating table 2": 0.06,
+    "Finished forward propagating table 3": 0.12,
+    "Finished forward propagating table 4": 0.2,
+    "Finished forward propagating table 5": 0.28,
+    "Finished forward propagating table 6": 0.36,
+    "Finished forward propagating table 7": 0.42,
+    "Finished prunning table 6": 0.43,
+    "Finished prunning table 5": 0.48,
+    "Finished prunning table 4": 0.51,
+    "Finished prunning table 3": 0.55,
+    "Finished prunning table 2": 0.58,
+    "Finished compressing tables 1 and 2": 0.66,
+    "Finished compressing tables 2 and 3": 0.73,
+    "Finished compressing tables 3 and 4": 0.79,
+    "Finished compressing tables 4 and 5": 0.85,
+    "Finished compressing tables 5 and 6": 0.92,
+    "Finished compressing tables 6 and 7": 0.98,
+}
+
+
+# https://kevinmccarthy.org/2016/07/25/streaming-subprocess-stdin-and-stdout-with-asyncio-in-python/
+async def _read_stream(stream, callback):
+    while True:
+        line = await stream.readline()
+        if line:
+            callback(line)
+        else:
+            break
+
+
+def parse_stdout(out):
+    out = out.rstrip()
+    print(out)
+    for k, v in progress.items():
+        if k in out:
+            print(f"Progress update: {v}")
+
+
+async def run(command):
+    process = await asyncio.create_subprocess_shell(
+        command, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE
+    )
+
+    await asyncio.wait(
+        [
+            _read_stream(
+                process.stdout,
+                lambda x: parse_stdout(x.decode("UTF8")),
+            ),
+            _read_stream(
+                process.stderr,
+                lambda x: print("STDERR: {}".format(x.decode("UTF8"))),
+            ),
+        ]
+    )
+
+    await process.wait()
+
+
+async def run_bladebit(args):
+    cmd = ""
+    for arg in args:
+        cmd += arg
+        cmd += " "
+    print(f"Running command: {cmd[:-1]}")
+    await run(cmd[:-1])
+
+
+def install_bladebit(root_path):
+    if is_bladebit_supported():
+        print("Installing dependencies.")
+        try:
+            subprocess.run(
+                [
+                    "sudo",
```

Comment:
It occurs to me that we'll have trouble with this step if plotting from the GUI starts the installation step

Reply:
There are utilities to handle GUI prompts if appropriate.  I can't say I know the 'proper' option at this point.  But yes, handling installation via other means would also be good.

Reply:
The issue is mostly resolved by having us package the prebuilt plotters for our installers, but there's still a question of how to handle this for users that clone the repo and build themselves. I'm leaning toward removing the option of installing the plotters on-demand when invoked from the GUI, but we could still build/install the plotters as part of install.sh and/or pull the prebuilt plotters. Basically anything that doesn't rely on sudo.

Reply:
I swear I don't mean to be nagging for installation from the GUI `:]`, if you are installing via a git clone then you are generally expected to be working from the CLI as well.  If you are using the package installers then certainly including prebuilt plotter binaries would be super nice.

But, it's not clear to me why the GUI itself is a major defining factor in this.  I think [`pkexec`](https://www.freedesktop.org/software/polkit/docs/latest/pkexec.1.html) is what I was thinking of for subprocessing with elevated privileges with a GUI password prompt.

Reply:
@altendky I didn't find a pkexec or gksudo option for macOS. Is there a cross-plat option that you're aware of? Could use AppleScript (via osascript) to handle macOS I suppose.

My knee-jerk reaction against handling sudo-based install from the GUI is based on a few of things...

1) Source-based installation is not going to be bulletproof. We'll have all sorts of OS versions attempting to build the plotters, with varying library support (I assume hardware support isn't as critical though). I view the build process as having a high probability of failure, and I doubt we'll want to take on ownership of ensuring that the plotters can be built across supported OS versions (beyond our own ability to build them for packaging).

2) The UX for supporting install could be as simple as clicking a button and just waiting several minutes without any feedback, but I think that would be a poor experience. Marginally better would be to show a spinner to at least communicate start/finished status, but that still strikes me as a poor experience. To give the user some indication of how far along installation is, and really, to communicate errors, I think we would need to pipe the output to a log and show in the GUI. This is just a bit more work than I think we've signed on for at the moment. And even then, I'm not sure there's a way to provide the user with actionable advice when a build failure occurs.

3) I'll admit that I'm viewing this through a macOS-centric lens, but having having Python processes invoke a admin privileges prompt is something I find to be highly suspicious. As a user, I'd feel more comfortable with a native helper app (code signed) that performs the installation, but this would be a fair amount of work to build/package.

That said, this is just how I personally feel about privileged script-based installation, but I'm flexible and open to being persuaded :)

Reply:
Nope, I'm not pushing for taking the time to make the GUI install work and build out that experience.  I agree with your comments.  Comfort with the CLI is likely needed to address many potential issues with the builds and that does reduce the value of investing in putting this into the GUI.

I didn't look hard but along with `osascript` I saw reference to `cocoasudo`.